framework/src/main/java/org/checkerframework/framework/util/dependenttypes/DependentTypesHelper.java

```diff
@@ -654,6 +656,61 @@ protected JavaExpression visitLocalVariable(
     convertAnnotatedTypeMirror(stringToJavaExpr, atm);
   }
 
+  /**
+   * Delocalizes dependent type annotations in {@code atm} so that they can be placed on the
+   * declaration of the given method or constructor being invoked. Used by whole program inference
+   * to infer dependent types for method/constructor parameters based on the actual arguments used
+   * at call sites.
+   *
+   * @param atm the annotated type mirror to delocalize
+   * @param invocationTree the method or constructor invocation
+   * @param arguments the actual arguments to the method or constructor
+   * @param methodElt the declaration of the method or constructor being invoked
+   */
+  public void delocalizeAtCallsite(
+      AnnotatedTypeMirror atm,
+      Tree invocationTree,
+      List<Node> arguments,
+      ExecutableElement methodElt) {
+
+    if (!hasDependentType(atm)) {
+      return;
+    }
+
+    // For use in stringToJavaExpr below, to avoid re-computation. Especially
+    // important for the TreePath, which is expensive to compute.
+    List<JavaExpression> argsAsExprs =
+        arguments.stream().map(LocalVariable::fromNode).collect(Collectors.toList());
+    TreePath path = factory.getPath(invocationTree);
+
+    StringToJavaExpression stringToJavaExpr =
+        stringExpr -> {
+          JavaExpression expr =
+              StringToJavaExpression.atPath(stringExpr, path, factory.getChecker());
+          JavaExpressionConverter jec =
+              new JavaExpressionConverter() {
+                @Override
+                public JavaExpression convert(JavaExpression javaExpr) {
+                  // if javaExpr is an argument to the method,
+                  // then return formal parameter expression.
+                  int index = argsAsExprs.indexOf(javaExpr);
```

Comment:
What about converting the receiver expression at the call site to `this`? Should `visitThisReference` also return null (unless the receiver expression is at the call site is `this`)?

Reply:
This hasn't been implemented.  For example, if the call sit is `expr.method(param)` and the type of `param` is  `@SameLen("expr")`, then it can be viewpoint-adapted to `@SameLen("this")`.  Maybe you don't want to implement that right now?  If so, you might to leave a TODO.

Reply:
Ah, I misunderstood what you were suggesting. WPI doesn't actually have support for inferring types on the receivers of methods from call sites, which is a TODO for WPI. I've added that TODO explicitly in the appropriate place in the WPI code with a note about the need to update `DependentTypesHelper` as well when that's implemented, and maybe I'll try implementing it later this week if I get a chance.

I also wrote a test for the situation you're describing, but commented it out (since it's still TODO).